### PR TITLE
Lazy load Room Receipts

### PR DESCRIPTION
**Pull request by ismailgulek**
Part of vector-im/element-ios#4383

### CODE REVIEW

File: MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.m [hunk 1]
```diff
@@ -168,40 +168,47 @@ - (NSString *)partialTextMessageOfRoom:(NSString *)roomId
...
+        
+        if (receiptsByUserId)
         {
-            for (NSString* userId in receiptsByUserId)
+            @synchronized (receiptsByUserId)
             {
-                MXReceiptData* receipt = receiptsByUserId[userId];
-
-                if (receipt && [receipt.eventId isEqualToString:eventId])
+                NSArray<MXReceiptData*> *receipts = [[receiptsByUserId allValues] filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"eventId == %@", eventId]];
```
Comment (manuroe) [inline MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.m:192]: Now that we have a completion block for this method, can we execute this block on a separated thread?
I remembered that operation was expensive.

Comment (manuroe) [reply MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.m:192]: Before doing that, please measure the duration of this call in MatrixHQ.

**Review by manuroe — COMMENTED** [review]
All good. I added another request while we are improving this part.

File: MatrixSDK/Data/Store/MXFileStore/MXFileStore.m [hunk 2]
```diff
@@ -830,6 +833,46 @@ - (MXMemoryRoomStore*)getOrCreateRoomStore:(NSString*)roomId
...
+    {
+        NSString *roomFile = [self readReceiptsFileForRoom:roomId forBackup:NO];
+        if ([[NSFileManager defaultManager] fileExistsAtPath:roomFile])
+        {
+            @try
+            {
+                NSDate *startDate = [NSDate date];
+                receiptsDictionary = [NSKeyedUnarchiver unarchiveObjectWithFile:roomFile];
+                receiptsByRoomId[roomId] = receiptsDictionary;
+                MXLogDebug(@"[MXFileStore] Loaded read receipts of room: %@ in %.0fms, in main thread: %@", roomId, [[NSDate date] timeIntervalSinceDate:startDate] * 1000, [NSThread isMainThread] ? @"YES" : @"NO");
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Data/Store/MXFileStore/MXFileStore.m]: Updated to log for only main thread as warning.

File: MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.m (hunk 1 above)
Comment (ismailgulek) [reply MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.m]: Looks negligible like under 1 ms. Anyway, i've moved that into another queue.

File: MatrixSDK/Data/Store/MXFileStore/MXFileStore.m [hunk 3]
```diff
@@ -830,6 +833,46 @@ - (MXMemoryRoomStore*)getOrCreateRoomStore:(NSString*)roomId
...
+            }
+            @catch (NSException *exception)
+            {
+                MXLogDebug(@"[MXFileStore] Warning: loadReceipts file for room %@ has been corrupted. Exception: %@", roomId, exception);
+                
+                // We used to reset the store and force a full initial sync but this makes the app
+                // start very slowly.
+                // So, avoid this reset by considering there is no read receipts for this room which
+                // is not probably true.
+                // TODO: Can we live with that?
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Data/Store/MXFileStore/MXFileStore.m:862]: I've copied this comment from `preloadReceipts` method. I think we can handle the _corrupted_ room and _initial room sync_ later.

**Review by manuroe — APPROVED** [review]

**Pull request merged**